### ISSUE TITLE
Parallel processing

### ISSUE THREAD

**Issue by schlegelp** [navis-org/navis#40]
Question for @clbarnes:

I recently added a `@map_neuronlist` [link](https://github.com/schlegelp/navis/blob/780280dfc8d23092e698b95aadf9c719d88dcd80/navis/utils/misc.py#L90) decorator that just applies the decorated function over all neurons in a neuron list. This saves a lot of boilerplate. 

My plan had been to extend this decorator to allow parallel processing. A mock example:

```Python
@map_neuronlist(desc='Pruning', allow_parallel=True)
def prune_neuron(x):
   # do stuff 
   return x
```

I tried that now by modifying the decorator to look something like this (again mock example):
```Python
def map_neuronlist(desc=None, allow_parallel=False):
    def decorator(function):
        @functools.wraps(function)
        def wrapper(nl, *args, **kwargs):
            if kwargs.pop('parallel', False) and allow_parallel:
                with multiprocessing.Pool() as pool:
                    # Map function over all neurons
                    res pool.imap(function, nl)
            else:
                res = [function(n) for n in nl]
            return NeuronList(res) 
        return wrapper    
```

Unfortunately, it does not work like that because pickling the decorated function fails with something along the lines of:
```Python
PicklingError: Can't pickle <function prune_by_strahler at 0x131b30710>: it's not the same object as navis.morpho.manipulation.prune_by_strahler
```

I found two ways to work around this:

1. Don't use `@decorators` and keep undecorated function. For the first example this would look something like this:

```Python
def _prune_neuron(x):
   # do stuff 
   return x

prune_neuron = map_neuronlist(desc='Pruning', allow_parallel=True)(x)
```

Not exactly pretty but it works. 

2. Use [`pathos`](https://github.com/uqfoundation/pathos) instead of the built-in `multiprocessing`. `pathos.multiprocessing.ProcessingPool` uses `dill` instead of `pickle` and seems to be able to deal with this issue. It's seems also much faster to spawn the processes but I haven't tested that properly. This would introduce a new (soft?) dependency.

Thoughts/suggestions?

**Comment by clbarnes** [comment]
Pathos looks good to me! Agree that it could be an optional extra - the boilerplate for throwing the error when the function is called instead of on import would only need to be written in the decorator definition.

I'm aware of dill but am not very familiar with it - with any luck it also uses out-of-band data like [pickle 5](https://www.python.org/dev/peps/pep-0574/) does. That massively speeds up the ser/deser of numpy arrays. The default pickle has been 4 since python 3.8, so I suspect that's what the multiprocessing queues use.

**Comment by schlegelp** [comment]
Actually your other suggestion already exists in `NeuronList.apply` :)

I was thinking of keeping that because it generalizes to any function - including non-navis functions. The advantage of the decorator is that we know that the output will be a NeuronList which makes some things easier.

**Comment by clbarnes** [comment]
Aha! My Ctrl+f was insufficient. Yes, that makes sense!

**Comment by schlegelp** [comment]
Implemented with 1c8ffe63c8ed764d49228874f983545c6feb86e9
